src/context/TimerContext.tsx

```diff
@@ -0,0 +1,205 @@
+                activeTimerIndex: nextIndex,
+                globalTimer: 0,
+            };
+        }
+
+        case 'RESET_TIMER_STATE':
+            return {
+                ...state,
+                timers: state.timers.map(resetTimer),
+                activeTimerIndex: null,
+                globalTimer: 0,
+                timerStatus: TimerStatus.READY,
+            };
+
+        case 'SET_TIME':
+            return { ...state, globalTimer: action.payload };
+
+        case 'COMPLETE_ALL': {
+            const completedWorkout = {
+                id: Date.now().toString(),
+                date: new Date().toISOString(),
+                totalDuration: calculateTotalWorkoutTime(state.timers),
+                timers: state.timers.map(timer => {
+                    // Calculate total duration dynamically for 'xy' and 'tabata' timers
+                    let totalDuration = timer.duration;
+
+                    if (timer.type === 'xy' && timer.workTime !== undefined) {
+                        totalDuration = (timer.rounds || 1) * timer.workTime;
+                    } else if (timer.type === 'tabata' && timer.workTime !== undefined && timer.restTime !== undefined) {
+                        totalDuration = (timer.rounds || 1) * (timer.workTime + timer.restTime);
+                    }
+
+                    return {
+                        id: timer.id,
+                        name: timer.name,
+                        type: timer.type,
+                        duration: totalDuration,
+                        rounds: timer.rounds,
+                        workTime: timer.workTime,
+                        restTime: timer.restTime,
+                    };
+                }),
+            };
+
+            // Save to localStorage
+            const history = JSON.parse(localStorage.getItem('workoutHistory') || '[]');
+            localStorage.setItem('workoutHistory', JSON.stringify([...history, completedWorkout]));
+
+            return { ...state, globalTimer: 0, activeTimerIndex: null, timerStatus: TimerStatus.COMPLETE };
+        }
+
+        default:
+            throw new Error(`Unhandled action type: ${(action as TimerAction).type}`);
+    }
+};
+
+// Context
+const TimerContext = createContext<{ state: TimerState; dispatch: React.Dispatch<TimerAction> } | null>(null);
+
+// Provider
+export const TimerProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
+    const [state, dispatch] = useReducer(timerReducer, initialState);
+
+    // Save state to local storage periodically
+    useEffect(() => {
+        const saveStateToLocalStorage = () => {
+            const { timers, timerStatus, activeTimerIndex, globalTimer } = state;
+            const stateToSave = { timers, timerStatus, activeTimerIndex, globalTimer, lastSavedAt: Date.now() };
+            localStorage.setItem('workoutState', JSON.stringify(stateToSave));
+        };
+
+        saveStateToLocalStorage(); // Save whenever state changes
```

Comment:
Wasn't able to test this because the URL state is not working to load. I tried refreshing the page and it clears the timer queue. Also tried the Share feature with the tinyurl but that didn't work either. I see the encrypted timers query param but its not loading it correctly.